x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/HasApplicationPrivilegesIT.java

```diff
@@ -151,6 +151,32 @@ public void testUserWithWildcardPrivileges() throws Exception {
         }
     }
 
+    public void testNamed() throws IOException {
+        createApplicationPrivilege("app", "write", new String[] { "action:write/*" });
+        createRole("correct", "app", new String[] { "read" }, new String[] { "*" });
+        createRole("wrong", "app", new String[] { "read", randomFrom("create", "write") }, new String[] { "*" });
+
+        var password = new SecureString("password-123");
+        createUser("correct", password, Set.of("correct"));
+        createUser("wrong", password, Set.of("wrong"));
+
+        {
+            var reqOptions = RequestOptions.DEFAULT.toBuilder()
+                .addHeader("Authorization", UsernamePasswordToken.basicAuthHeaderValue("correct", password))
+                .build();
+            var actual = hasPrivilege(reqOptions, "app", new String[] { "read" }, new String[] { "resource" });
+            assertSinglePrivilege(actual, "resource", "read", true);
+        }
+
+        {
+            var reqOptions = RequestOptions.DEFAULT.toBuilder()
+                .addHeader("Authorization", UsernamePasswordToken.basicAuthHeaderValue("wrong", password))
+                .build();
+            var actual = hasPrivilege(reqOptions, "app", new String[] { "read" }, new String[] { "resource" });
+            assertSinglePrivilege(actual, "resource", "read", true);
```

Comment:
This fails.

Reply:
This is the problem: https://github.com/elastic/elasticsearch/blob/79f801b7547a44c9b41086e85d586a7502da95af/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java#L194

We are checking for equality, but when there's more than one privilege on the Role, we fail this check since name is `Set.of("create", "read")` for the privilege on the role, and it does not equal  `Set.of("read")` -- the privilege from the request.